### DISCUSSION TITLE
Azure OAuth CSRF State Not Equal Error

### DISCUSSION THREAD

**Discussion by ahipp13** [apache/airflow#28098] [General] [answered]
### Apache Airflow version

2.4.3

### What happened

We have enabled Microsoft Azure OAuth for our Airflow implementation. When we try to log in, we get a CSRF error: 

[2022-11-28 22:04:58,744] {views.py:659} ERROR - Error authorizing OAuth access token: mismatching_state: CSRF Warning! State not equal in request and response.                                                                 ││ airflow-web [2022-11-28 22:04:58,744] {views.py:659} ERROR - Error authorizing OAuth access token: mismatching_state: CSRF Warning! State not equal in request and response.

We have taken a look at both the sending and receiving URLs in the browser and the state is the exact same. Down below are pictures of the state for the request and response:
![2022-11-28_16-22-46](https://user-images.githubusercontent.com/118911990/204394405-2e7a0029-9b18-4090-89e6-52a5a41dc25d.png)
![2022-11-28_16-26-24](https://user-images.githubusercontent.com/118911990/204394418-d2603bbf-6668-4fda-ad75-58fce46c3d44.png)

### What you think should happen instead

We should be able to log into our Airflow application. We had the exact same setup using Airflow 2.2.5 and everything worked just fine. 

### How to reproduce

Down below is a copy of our webserver_config.py. We are currently running Airflow 2.4.3 on Kubernetes with the Airflow Community helm chart version 8.6.1 (located here: https://github.com/airflow-helm/charts). We are also using a postgres external database as our metadata db. 

```
from flask_appbuilder.security.manager import AUTH_OAUTH
from airflow.www.security import AirflowSecurityManager
import logging
from typing import Dict, Any, List, Union
import os
import sys

#Add this as a module to pythons path
sys.path.append('/opt/airflow')

log = logging.getLogger(__name__)
log.setLevel(os.getenv("AIRFLOW__LOGGING__FAB_LOGGING_LEVEL", "DEBUG"))

class AzureCustomSecurity(AirflowSecurityManager):
    # In this example, the oauth provider == 'azure'.
    # If you ever want to support other providers, see how it is done here:
    # https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/manager.py#L550
    def get_oauth_user_info(self, provider, resp):
        # Creates the user info payload from Azure.
        # The user previously allowed your app to act on their behalf,
        #   so now we can query the user and teams endpoints for their data.
        # Username and team membership are added to the payload and returned to FAB.
        if provider == "azure":
            log.debug("Azure response received : {0}".format(resp))
            id_token = resp["id_token"]
            log.debug(str(id_token))
            me = self._azure_jwt_token_parse(id_token)
            log.debug("Parse JWT token : {0}".format(me))
            return {
                "name": me.get("name", ""),
                "email": me["upn"],
                "first_name": me.get("given_name", ""),
                "last_name": me.get("family_name", ""),
                "id": me["oid"],
                "username": me["oid"],
                "role_keys": me.get("roles", []),
            }

# Adding this in because if not the redirect url will start with http and we want https
os.environ["AIRFLOW__WEBSERVER__ENABLE_PROXY_FIX"] = "True"
WTF_CSRF_ENABLED = False
CSRF_ENABLED = False
AUTH_TYPE = AUTH_OAUTH
AUTH_ROLES_SYNC_AT_LOGIN = True  # Checks roles on every login
# Make sure to replace this with the path to your security manager class
FAB_SECURITY_MANAGER_CLASS = "webserver_config.AzureCustomSecurity"
# a mapping from the values of `userinfo["role_keys"]` to a list of FAB roles
AUTH_ROLES_MAPPING = {
    "airflow_dev_admin": ["Admin"],
    "airflow_dev_op": ["Op"],
    "airflow_dev_user": ["User"],
    "airflow_dev_viewer": ["Viewer"]
    }
# force users to re-auth after 30min of inactivity (to keep roles in sync)
PERMANENT_SESSION_LIFETIME = 1800
# If you wish, you can add multiple OAuth providers.
OAUTH_PROVIDERS = [
    {
        "name": "azure",
        "icon": "fa-windows",
        "token_key": "access_token",
        "remote_app": {
            "client_id": "CLIENT_ID",
            "client_secret": 'AZURE_DEV_CLIENT_SECRET',
            "api_base_url": "https://login.microsoftonline.com/TENANT_ID",
            "request_token_url": None,
            'request_token_params': {
                'scope': 'openid email profile'
            },
            "access_token_url": "https://login.microsoftonline.com/TENANT_ID/oauth2/v2.0/token",
            "access_token_params": {
                'scope': 'openid email profile'
            },
            "authorize_url": "https://login.microsoftonline.com/TENANT_ID/oauth2/v2.0/authorize",
            "authorize_params": {
                'scope': 'openid email profile',
            },
            'jwks_uri':'https://login.microsoftonline.com/common/discovery/v2.0/keys',
        },
    },
]
```

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==6.0.0
apache-airflow-providers-celery==3.0.0
apache-airflow-providers-cncf-kubernetes==4.4.0
apache-airflow-providers-common-sql==1.2.0
apache-airflow-providers-docker==3.2.0
apache-airflow-providers-elasticsearch==4.2.1
apache-airflow-providers-ftp==3.1.0
apache-airflow-providers-google==8.4.0
apache-airflow-providers-grpc==3.0.0
apache-airflow-providers-hashicorp==3.1.0
apache-airflow-providers-http==4.0.0
apache-airflow-providers-imap==3.0.0
apache-airflow-providers-microsoft-azure==4.3.0
apache-airflow-providers-mysql==3.2.1
apache-airflow-providers-odbc==3.1.2
apache-airflow-providers-postgres==5.2.2
apache-airflow-providers-redis==3.0.0
apache-airflow-providers-sendgrid==3.0.0
apache-airflow-providers-sftp==4.1.0
apache-airflow-providers-slack==6.0.0
apache-airflow-providers-sqlite==3.2.1
apache-airflow-providers-ssh==3.2.0

### Deployment

Other 3rd-party Helm chart

### Deployment details

We are currently running Airflow 2.4.3 on Kubernetes with the Airflow Community helm chart version 8.6.1 (located here: https://github.com/airflow-helm/charts). We are also using a postgres external database as our metadata db. 

### Anything else

This problem occurs every time we try to log into the Airflow Webserver using Azure OAuth.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
I think this is at most a discussion, not Airflow issue. 

There are I think several ways it can happen:

* you can have different webserver_keys in your webservers 
* some libraries needs changing (similar answres here https://stackoverflow.com/questions/61922045/mismatchingstateerror-mismatching-state-csrf-warning-state-not-equal-in-reque) 
* It could be an authlib issue: https://github.com/lepture/authlib/issues/376 (but I see we found similar issue and you reported it there)

You can also try to check in FAB https://github.com/dpgaspar/Flask-AppBuilder - this is where the authentication and Oauth in Airflow comes from. They have definitely more expertise in the area than we have and will be able to help you more efficienty.

I am afraid that's as much of a help you can get for now - unless somene else who solved similar issue in the path can come up with.

**Reply by ahipp13** [reply]
We are only running 1 webserver, so I do not think it has anything to do with the webserver_keys. As far as the libraries go, I tried to downgrade to the authlib version said in that stack overflow, and it still did not work. Maybe I could try downgrading all of Flask to the versions they were in 2.2.5 as well and try it?

I guess I will go try and make posts in Flask and the Community Helm Chart and see if anybody else has any insights. I just can't believe I cannot find anybody else having the same issue... Thanks for your response, will update this if I find anything

**Reply by ahipp13** [reply]
Trying to downgrade libraries did not work, just keep getting other errors. Have posted in FAB and the community helm chart but no helpful responses yet. I went back and tried to create the webserver_config by scratch, going through the documentation carefully. Still, the same CSRF state error :/ Pretty stumped at this point, will keep this updated if a solution is found.

**Reply by ahipp13** [reply]
@potiuk Can you provide me the location of where Airflow sets up the Flask app and sessions?

To give some context, I have did some painful debugging, and have come to find out that for some reason within authlib it is returning the state as None when in the urls the state is coming up. If you want to read more about my further debugging, you can go to here: https://github.com/lepture/authlib/issues/518

but the guy at authlib basically told me to 1) Make sure the Flask app has set secure_key and 2) Check if the session works, so I just need to know where Airflow sets up all this stuff. Thanks in advance!

**Reply by potiuk** [reply]
Pretty natural place is `airflow/www/app.py`. It has `def create_app(config=None, testing=False)`

**Reply by potiuk** [reply]
"""Create a new instance of Airflow WWW app""" is the javadoc so looks like it.

**Reply by ahipp13** [reply]
@potiuk to give an update, it looks like there is something wrong with the session. When trying to access it within Authlib, it looks like a created session: <SqlAlchemySession {'_permanent': True, '_fresh': False}> but when you try and get any value (like print(session.get("name"))), it returns as "None".  So there seems to be problem with the session. If you want a more detailed update you can click the link above that links to the issue within authlib.

**Reply by potiuk** [reply]
I guess since it is open in  FAB - we have to wait for what Daniel is going to tell.

**Reply by potiuk** [reply]
For the record here https://github.com/dpgaspar/Flask-AppBuilder/issues/1957

**Reply by ahipp13** [reply]
While I know this isn't the spot to get this fixed, I have did some more debugging and have more information on what is going on that I am going to share here so everybody has the same information.

Looks like I figured out what is happening but am unsure on how to fix. I have put in a ton of debugging statements to follow the flow of code to see what was happening in the background. But, in interest of keeping this update short, I will just summarize what I found and what I think is happening.
 
We start in flask_appbuilder/security/views.py in the "def login(self, provider: Optional[str] = None) -> WerkzeugResponse:" function under the "AuthOAuthView(AuthView) class. This function returns a call to another function, "authorize_redirect".
 
The function" authorize_redirect" is found in authlib/integrations/flask_client/apps.py at line 39. This function does a lot of things, but the most important is at the end when it calls "save_authorize_data". "save_authorize_data" is in the same file, and found at line 32. This function calls a function called "set_state_data". This leads us to authlib/integrations/base_client/framework_integration.py, where it sets the session data and returns.
 
Before the function "Authorize_redirect" returns back to "views.py", I printed out the session to see what it looked like. This is what is printed:
![image](https://user-images.githubusercontent.com/118911990/209407383-b004fa9b-d9c8-4c2b-8486-9944bcc1b361.png)

Clearly, you can see here that session has the correct state. So it returns the redirect url, and according to this in the return statement of the "login" function: "redirect_uri=url_for( ".oauth_authorized", provider=provider _external=True)" my guess is that the next place its going is the "oauth_authorized" function within the "views.py" file. So, I put print statements at the top of this file to print out the session, and this is what I get:
![image](https://user-images.githubusercontent.com/118911990/209407411-abbfab7a-846c-48e5-b854-cd8896244c04.png)

So, somehow the redirect is causing the session to lose all its data…

**Reply by ahipp13** [reply]
@potiuk More updates on this issue. I got some debugging advice from the guy from authlib. he said to do this:

you can debug your session issue with:

@app.route('/set-session')
def set_session():
    session['test'] = 'foo'
    return 'set session'

@app.route('/read-session')
def read_session():
    value = session.get('test', '')
    return value

1. visit /set-session to set the session
2. open another browser tab, visit /read-session to see if you can get the value



This is what I did to get this to work: 



What I had to do was edit the views.py file in the airflow source code (airflow/www/views.py). This is what I put in (the class starts at line 561):

```
class Airflow(AirflowBaseView):
    """Main Airflow application."""
    
    @expose('/set-session')
    def set_session(self):
        flask_session['test'] = 'foo'
        return 'set session'

    @expose('/read-session')
    def read_session(self):
        value = flask_session.get('test','')
        return value   
```

I then redeployed my application. I went to the 2 urls, and this is the output:

![image](https://user-images.githubusercontent.com/118911990/211388573-dca084c1-0c5d-4588-86f1-ea8364879e3c.png)

![image](https://user-images.githubusercontent.com/118911990/211388804-0ae18ae8-cbcf-4aad-bee3-5e7fa1997fa5.png)

As you can see, in the other browser I was unable to get the value. According to the authlib guy, this means that the application's session is not configured properly. So my question is where does Airflow configure the session? Again this is all being posted in this issue: https://github.com/lepture/authlib/issues/518

**Comment by potiuk** [comment]
Another option is - yoy might check in the community chart - it's a project that is not run by the Apache Airflow Community - it's run independently and maybe there you will find some better help.

**Comment by ahipp13** [comment]
@potiuk I have finally found what the source of this problem was, as well as a solution.
 
Doing more debugging, I found that the problem with the session was coming from the webserver_config.py file that I had created. So, what I did was started with as barebones of a Oauth Webserver_config.py file as I could, and kept adding lines until one of them screwed up the session. In doing this I found that it was this line: 

PERMANENT_SESSION_LIFETIME = 1800
 
That was causing the session to not persist. I am not too sure as to why. The reason I had put this in my webserver_config.py file in the first place is because it is in the FAB documentation(https://flask-appbuilder.readthedocs.io/en/latest/security.html#), but in further researching I found that it is already being set in the creation of the flask app by Airflow and is an configuration option: AIRFLOW__WEBSERVER__SESSION_LIFETIME_MINUTES
 
Now why this worked in Airflow version 2.2.5 and not in Airflow 2.4.3, I am not for sure.
 
Also through my debugging I found out that FAB natively supports Azure now, so a custom security class and user info handler function is not needed.
 
So, to answer this question, the solution to my problem was to start using thie webserver_config.py file down below. 
```
from __future__ import annotations
import os
from airflow.www.fab_security.manager import AUTH_OAUTH
# Default Auth Type
# from airflow.www.fab_security.manager import AUTH_DB

basedir = os.path.abspath(os.path.dirname(__file__))

# Flask-WTF flag for CSRF
WTF_CSRF_ENABLED = True

# ----------------------------------------------------
# AUTHENTICATION CONFIG
# ----------------------------------------------------
# For details on how to set up each of the following authentication, see
# http://flask-appbuilder.readthedocs.io/en/latest/security.html# authentication-methods
# for details.

# The authentication type
AUTH_TYPE = AUTH_OAUTH

# registration configs
AUTH_USER_REGISTRATION = True  # allow users who are not already in the FAB DB
AUTH_USER_REGISTRATION_ROLE = "Public"  # this role will be given in addition to any AUTH_ROLES_MAPPING

# Specifying Oauth Providers
OAUTH_PROVIDERS = [
    {
        "name": "azure",
        "icon": "fa-windows",
        "token_key": "access_token",
        "remote_app": {
            "client_id": "APPLICATION_CLIENT_ID",
            "client_secret": "AZURE_DEV_CLIENT_SECRET",
            "api_base_url": "https://login.microsoftonline.com/AZURE_TENANT_ID/oauth2",
            "client_kwargs": {
                "scope": "User.read name preferred_username email profile upn",
                "resource": "APPLICATION_CLIENT_ID",
            },
            "request_token_url": None,
            "access_token_url": "https://login.microsoftonline.com/AZURE_TENANT_ID/oauth2/token",
            "authorize_url": "https://login.microsoftonline.com/AZURE_TENANT_ID/oauth2/authorize",
            'jwks_uri':'https://login.microsoftonline.com/common/discovery/v2.0/keys',
        },
    },
]

# Adding this in because if not the redirect url will start with http and we want https
os.environ["AIRFLOW__WEBSERVER__ENABLE_PROXY_FIX"] = "True"

#This maps our roles set in Azure to the roles in Airflow
AUTH_ROLES_MAPPING = {
    "airflow_dev_admin": ["Admin"],
    "airflow_dev_op": ["Op"],
    "airflow_dev_user": ["User"],
    "airflow_dev_viewer": ["Viewer"]
    }

# # if we should replace ALL the user's roles each login, or only on registration
AUTH_ROLES_SYNC_AT_LOGIN = True
```

**Comment by ahipp13** [comment]
@potiuk interesting note I want to add that should be reported as bug either on your side or FABs side. Looks like the problem is with the PERMANENT_SESSION_LIFETIME variable in general. I just tried to set this value to 30 minutes in my helm chart like this:

AIRFLOW__WEBSERVER__SESSION_LIFETIME_MINUTES: "30"

And the CSRF state error instantly came back. After taking this out, it worked again. So this should be reported as a bug.

**Reply by potiuk** [reply]
I am not sure it makes sense because we just merged #28730 which changes CSRF token to expire with session - which I think should fix the problem entirely.

**Reply by ahipp13** [reply]
I will take a look at that tomorrow

**Comment by ahipp13** [comment, marked as answer]
To add some context on this discussion, when I upgraded from Airflow 2.2.5 to 2.4.3, I was having problems. The first problem is listed here: https://github.com/apache/airflow/discussions/28099 I found a workaround to this problem, and when I put that workaround in place, it then lead me to this CSRF error.

I have went back and further found a better work around for discussion 28099 that not only takes away that issue, but it also takes away this issue as well.

So if you are having both integrityerrors and CSRF errors, look at discussion 28099 to see if that workaround fixes it. If you are just having CSRF errors and you are setting AIRFLOW__WEBSERVER__SESSION_LIFETIME_MINUTES, you can try and not set it and see if it works.
